File: SearchAlgorithms/DepthFirstSearch.java
import java.util.*;

public class DepthFirstSearch {

	DepthFirstSearch() {}

	public void search(List<List<Integer>> graph, boolean[] visited, int node) {
		visited[node] = true;
		System.out.println(node);

		for (int neighbor: graph.get(node)) {
			if (!visited[neighbor]) {
				search(graph, visited, neighbor);
			}
		}
	}
	
	public static void main(String[] args) {

		DepthFirstSearch depthFirstSearch = new DepthFirstSearch();
		
		int n = 5;
		List<List<Integer>> graph = new ArrayList<>();
		
		for (int i = 0; i < n; i++) {
			graph.add(new ArrayList<>());
		}

		graph.get(0).add(1);
        graph.get(0).add(2);
        graph.get(1).add(3);
        graph.get(1).add(4);

        boolean[] visited = new boolean[n];
        depthFirstSearch.search(graph, visited, 0);
		
	}
}

File: SearchAlgorithms/DepthFirstSearch.cpp
#include <iostream>
#include <vector> 

using namespace std;

class DepthFirstSearch {
	
	public:
		void dps(vector<vector<int>> vecz, bool visitedNodes[], int start, int target) {
		
			visitedNodes[start] = true;
			cout << "Visited node: " << start << endl;

			for (size_t i = 0; i <vecz[start].size() ; i++) {

				if (!visitedNodes[vecz[start][i]]) {
					dps(vecz, visitedNodes,vecz[start][i] , target);
				}					
			}
			
		} 
						
};

int main() {

	vector<vector<int>> vecz(7);

	vecz[0].push_back(1);
	vecz[0].push_back(2);
	vecz[1].push_back(3);
	vecz[1].push_back(4);
	vecz[2].push_back(5);
	vecz[2].push_back(6);
			
	bool visitedNodes[7];		
	int start = 0;
	int target = 4;
	
	DepthFirstSearch deptheFirstSearch;

	deptheFirstSearch.dps(vecz, visitedNodes, start, target);
}
